src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -351,3 +367,419 @@ deriving stock instance Read SizeRatio
+        Just version -> pure version
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding: SnapshotMetaData
+-------------------------------------------------------------------------------}
+
+-- SnapshotMetaData
+
+instance Encode SnapshotMetaData where
+  encode (SnapshotMetaData label tableType config levels) =
+         encodeListLen 4
+      <> encode label
+      <> encode tableType
+      <> encode config
+      <> encode levels
+
+instance DecodeVersioned SnapshotMetaData where
+  decodeVersioned ver@V0_0 = do
+      _ <- decodeListLenOf 4
+      SnapshotMetaData
+        <$> decodeVersioned ver <*> decodeVersioned ver
+        <*> decodeVersioned ver <*> decodeVersioned ver
+
+-- SnapshotLabel
+
+instance Encode SnapshotLabel where
+  encode (SnapshotLabel s) = encodeString s
+
+instance DecodeVersioned SnapshotLabel where
+  decodeVersioned V0_0 = SnapshotLabel <$> decodeString
+
+-- TableType
+
+instance Encode SnapshotTableType where
+  encode SnapNormalTable   = encodeListLen 1 <> encodeWord 0
+  encode SnapMonoidalTable = encodeListLen 1 <> encodeWord 1
+
+instance DecodeVersioned SnapshotTableType where
+  decodeVersioned V0_0 = do
+      _ <- decodeListLenOf 1
+      tag <- decodeWord
+      case tag of
+        0 -> pure SnapNormalTable
+        1 -> pure SnapMonoidalTable
+        _ -> fail ("[SnapshotTableType] Unexpected tag: " <> show tag)
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding: TableConfig
+-------------------------------------------------------------------------------}
+
+-- TableConfig
+
+instance Encode TableConfig where
+  encode config =
+         encodeListLen 7
+      <> encode mergePolicy
+      <> encode sizeRatio
+      <> encode writeBufferAlloc
+      <> encode bloomFilterAlloc
+      <> encode fencePointerIndex
+      <> encode diskCachePolicy
+      <> encode mergeSchedule
+    where
+      TableConfig
+        mergePolicy
+        sizeRatio
+        writeBufferAlloc
+        bloomFilterAlloc
+        fencePointerIndex
+        diskCachePolicy
+        mergeSchedule
+        = config
+
+instance DecodeVersioned TableConfig where
+  decodeVersioned v@V0_0 = do
+      _ <- decodeListLenOf 7
+      TableConfig
+        <$> decodeVersioned v <*> decodeVersioned v <*> decodeVersioned v
+        <*> decodeVersioned v <*> decodeVersioned v <*> decodeVersioned v
+        <*> decodeVersioned v
+
+-- MergePolicy
+
+instance Encode MergePolicy where
+  encode MergePolicyLazyLevelling =
+      encodeListLen 1 <> encodeWord 0
+
+instance DecodeVersioned MergePolicy where
+  decodeVersioned V0_0 =  do
+      _ <- decodeListLenOf 1
+      tag <- decodeWord
+      case tag of
+        0 -> pure MergePolicyLazyLevelling
+        _ -> fail ("[MergePolicy] Unexpected tag: " <> show tag)
+
+-- SizeRatio
+
+instance Encode SizeRatio where
+  encode Four = encodeWord64 4
```

Comment:
No need to be so specific with the size. CBOR doesn't use fixed width representations, so you don't need to "reserve" 64bits now just in case. It'll encode to the minimum representation size anyway. You can just use `encodeInt` / `decodeInt`. If for some reason in future we needed a bigger number type, we can switch and still read the old encoded data.